include/dlaf/communication/sync/reduce.h

```diff
@@ -13,24 +13,90 @@
 /// @file
 
 #include <mpi.h>
+#include "dlaf/common/data_descriptor.h"
 #include "dlaf/communication/communicator.h"
 #include "dlaf/communication/message.h"
 
 namespace dlaf {
 namespace comm {
 namespace sync {
 
-/// MPI_Reduce wrapper
+namespace internal {
+namespace reduce {
+
+/// @brief MPI_Reduce wrapper (sender side)
+/// MPI Reduce(see MPI documentation for additional info)
+/// @param reduce_operation MPI_Op to perform on @p input data coming from ranks in @p communicator
+template <class DataIn, class DataOut>
+void collector(Communicator& communicator, MPI_Op reduce_operation, DataIn input, DataOut output) {
```

Comment:
Consider a fast return if the communicator size is one.

Reply:
Good point!

I'm wondering:
- what it could be better to do in that case? Is it legit to reduce on a single rank? Or do we want to consider it an error?
- is it worth to add a test?

Reply:
- Yes it is legit.
- yes, since it is a special case with its implementation.

Reply:
Changes highlights:
- Test just with const input to reduce (it was not strictly redundant, but it may be an overkill)
- Implemented wayout for single rank communicator (just copy)
- Add debug check for validity of `rank_root` parameter
- Add const correctness for Data parameters (reference semantic, it is not const the value but the Data concept)
- Add `constexpr` to `dlaf_test::TypeUtilities`

Since you already reviewed this PR, I suggest you to give a look at the changes from your last review for all files with an exception for the file containing tests, for which it will be easier to look at all changes made from the beginning.